### PR TITLE
[#118865569] Fix bug of wrong Broadcast with Window Function

### PR DESCRIPTION
**Pull request by oarap**
When there is a Gather on top of the window function
(CPhysicalSequenceProject), there cannot be a broadcast below that
operator.

The fix got two parts. First, fix-up the CDistributionSpecAny to
remember the EOperatorId of who requests it. This will help the
reasoning later on in the CPhysicalSequenceProject to decide what kind
of required distribution spec to request from its child.

Second, we fix up the CPhysicalSequenceProject to avoid generating a
CDistributionReplicated if a Gather motion generated an
CDistributionSpecAny.

Besides those, we also cleanup the testing framework:
- rename Fst --> Actual and Snd --> Expected in UnitTest.
  This helped a lot reading what's expected and what actual plan
  produced in the gporca unit tests.
- refactored the CICGTest a little to separate window function tests out
- fix minidump option `-d` to have same configuration as `-U`

### CODE REVIEW

File: libgpopt/src/operators/CPhysicalSequenceProject.cpp [hunk 1]
```diff
@@ -355,6 +359,8 @@ CPhysicalSequenceProject::PdsRequired
 		return GPOS_NEW(pmp) CDistributionSpecReplicated();
 	}
 
+	// if the window operator has a partition by clause, then always ask
+	// request hashed distribution on the partition column
```
Comment (vraghavan78) [inline libgpopt/src/operators/CPhysicalSequenceProject.cpp]: remove ask?

File: libgpopt/src/operators/CPhysicalSequenceProject.cpp [hunk 2]
```diff
@@ -366,7 +372,23 @@ CPhysicalSequenceProject::PdsRequired
 		return GPOS_NEW(pmp) CDistributionSpecSingleton(CDistributionSpecSingleton::EstMaster);
 	}
 
-	return GPOS_NEW(pmp) CDistributionSpecReplicated();
+	if (CDistributionSpec::EdtAny == pdsRequired->Edt())
```
Comment (vraghavan78) [inline libgpopt/src/operators/CPhysicalSequenceProject.cpp]: Please add a comment about what this is trying to accomplish.

File: libgpopt/src/operators/CPhysicalSequenceProject.cpp [hunk 3]
```diff
@@ -366,7 +372,23 @@ CPhysicalSequenceProject::PdsRequired
...
+	{
+		CDistributionSpecAny *anySpec = CDistributionSpecAny::PdsConvert(pdsRequired);
+		if (COperator::EopPhysicalMotionGather != anySpec->GetRequestedOperatorId()
+				&& COperator::EopPhysicalFilter != anySpec->GetRequestedOperatorId())
+		{
```
Comment (vraghavan78) [inline libgpopt/src/operators/CPhysicalSequenceProject.cpp]: why Filter?

File: server/src/unittest/gpopt/minidump/CICGTest.cpp [hunk 4]
```diff
@@ -44,6 +44,21 @@ ULONG CICGTest::m_ulTestCounterPreferHashJoinToIndexJoin = 0;
...
+		"../data/dxl/minidump/WinFunc-Simple.mdp",
+		"../data/dxl/minidump/WindowFrame-SingleEdged.mdp",
+		"../data/dxl/minidump/Join-WinFunc-Preds.mdp",
+		"../data/dxl/minidump/Lead-Lag-WinFuncs.mdp",
+		"../data/dxl/minidump/Preds-Over-WinFunc1.mdp",
+		"../data/dxl/minidump/Preds-Over-WinFunc2.mdp",
+		"../data/dxl/minidump/Preds-Over-WinFunc3.mdp",
+		"../data/dxl/minidump/Preds-Over-WinFunc4.mdp",
+		"../data/dxl/minidump/Preds-Over-WinFunc5.mdp",
+};
```
Comment (vraghavan78) [inline server/src/unittest/gpopt/minidump/CICGTest.cpp]: Can we make another test like CWindowTest?

File: libgpopt/src/operators/CPhysicalSequenceProject.cpp (hunk 3 above)
Comment (oarap) [inline libgpopt/src/operators/CPhysicalSequenceProject.cpp]: I think it is just to make the test case to pass.

File: server/src/unittest/gpopt/minidump/CWindowTest.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+	return CTestUtils::EresUnittest_RunTestsWithoutAdditionalTraceFlags
+				(
+					rgszWindowFunctionFileNames,
+					&ulTestCounter,
+					GPOS_ARRAY_SIZE(rgszWindowFunctionFileNames),
+					true,
+					true
+				);
+	/*
```
Comment (vraghavan78) [inline server/src/unittest/gpopt/minidump/CWindowTest.cpp]: Please remove commented code.

File: libgpopt/src/operators/CPhysicalSequenceProject.cpp [hunk 6]
```diff
@@ -366,6 +372,20 @@ CPhysicalSequenceProject::PdsRequired
 		return GPOS_NEW(pmp) CDistributionSpecSingleton(CDistributionSpecSingleton::EstMaster);
 	}
 
+	// if the required distribution type is ANY then we check which
+	// physical operators are generating ANY distribtion and if the ANY
+	// is requested by MotionGather or Filter, we create Singleton.
```
Comment (vraghavan78) [inline libgpopt/src/operators/CPhysicalSequenceProject.cpp]: You do not explain why we special case for GatherMotion and Filter?

File: data/dxl/minidump/NoBroadcastUnderGatherForWindowFunction.mdp [hunk 7]
```diff
@@ -0,0 +1,603 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
```
Comment (vraghavan78) [inline data/dxl/minidump/NoBroadcastUnderGatherForWindowFunction.mdp]: DDL and Query?

**Comment by vraghavan78** [comment]
Please elaborate the comment. Else LGTM

File: server/src/unittest/gpopt/minidump/CWindowTest.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+//	Greenplum Database
+//	Copyright (C) 2016 Pivotal Software
+//
+//	@filename:
+//		CWindowTest.cpp
+//
+//	@doc:
+//		Test for Window functions
+//
+//	@owner:
```
Comment (d) [inline server/src/unittest/gpopt/minidump/CWindowTest.cpp]: a bit of bikeshedding, i hate `@owner` and `@test`, they are blindly copy-pasted, and are totally irrelevant

**Comment by xinzweb** [comment]
+1

**Pull request closed without merging**